### PR TITLE
daemon mount namespaces

### PR DESCRIPTION
**Pull request by vbatts**
This change covers two of the contrib init's. The daemon ought to be invoked in its own mount namespace, unshared from the root mount namespace.

Some of the error most often seen affect the devicemapper graph driver, seen as EBUSY or "device or resource busy".
Though others are likely to run in to this as folks will have containers that will have mounts or even devices being mounted. 

If any other pid outside of the docker daemon unshares their mount namespace, and that  pid's mountinfo includes a mount reference of a container, then even when the container exits, and attempts to remove, there kernel will return that the path is busy, due to the mount reference by the outside pid.

I've written about this here: http://blog.hashbangbash.com/2014/11/docker-devicemapper-fix-for-device-or-resource-busy-ebusy/

I did not include a commit for the init scripts with `start-stop-daemon`, as I would like them to get more testing first. See: 
- openrc - http://pastebin.com/tT1jUhzR
- sysvinit-debian - http://pastebin.com/7hbkSTna

(and for completeness, here is the change for ubuntu's upstart http://pastebin.com/D5MsVCUK)

### CODE REVIEW

**Comment by tianon** [comment]
LGTM, although I think we should wait until after 1.5 (and then include the `start-stop-daemon` and upstart changes too)

**Comment by tianon** [comment]
Unless you can make a compelling argument for this not breaking other stuff so close to our freeze window. :wink:

**Comment by vbatts** [comment]
https://github.com/docker/docker/issues?q=is%3Aissue+is%3Aopen+%22device+or+resource+busy%22 bothers me

**Comment by vbatts** [comment]
and, this is ./contrib/ ;-)

**Comment by thaJeztah** [comment]
^^ Add a line to those files: "In case of emergency, call Vincent +1 555 ...", then it should be good to merge :smile_cat:

**Comment by vbatts** [comment]
@thaJeztah woah there, cowboy ;-)

**Comment by thaJeztah** [comment]
I'm not very familiar with this stuff, but would this also have a positive effect on AUFS not freeing inodes I saw reported?

**Comment by vbatts** [comment]
@thaJeztah do you have a link to that conversation?

**Comment by thaJeztah** [comment]
Yup, read it earlier today, that's why I thought of it; https://github.com/docker/docker/issues/9755#issuecomment-70716037

**Comment by dqminh** [comment]
@vbatts Do you think this will fix aufs `device and resource busy` bugs too ? I had a few reports about it recently, and the symptoms look familiar.

**Comment by tianon** [comment]
@vbatts For the redhat sysvinit change, I agree 100% that the only repercussions would be to the redhat packagers who are quite familiar with creating out-of-band patches.

For the systemd change, this is used in the packages we release, so I'm much more leery (since a problem there would mean we have to re-release depending on the severity and the number of times it gets reported :disappointed:).

Regardless, we'll need our systemd maintainers to weigh in: @lsm5 @philips @jfrazelle

**Comment by tianon** [comment]
(see also https://github.com/docker/docker/blob/9de604ad68033d08721ce61db7300cee338837e2/project/make/ubuntu#L27-L47)

**Comment by cpuguy83** [comment]
Didn't @crosbymichael already try doing this, although within Docker itself? #4599

Also, what happens in situations where someone has mounted an fs on the host and wants to then bind-mount that in a container?

**Comment by unclejack** [comment]
@cpuguy83 No, that wasn't merged because there were other problems to take into account.

**Comment by vbatts** [comment]
@tianon well, fedora and centos7 had been using `MountFlags=private` for a number of versions. So it has had exposure and testing. They have recently switched to `MountFlags=slave`, for the purpose of ensuring the shared-subtree mounts propagate in the right direction.

**Comment by tianon** [comment]
@vbatts OK, that's fair; so, if the `contrib/init/systemd` maintainers (@lsm5 @philips @jfrazelle) sign-off on this, I'm OK to merge it for 1.5

I think it's probably prudent that we at least ping @maxamillion for the sysvinit change. :smile:

Also, for future reference, the argument that "it's just `contrib`" doesn't hold any water with me.  We use bits of `contrib` in our releases (as I linked above).

**Comment by vbatts** [comment]
@tianon fair. I like to keep contrib fresh and current too. :heart:

**Comment by jfrazelle** [comment]
LGTM

File: contrib/init/systemd/docker.service [hunk 1]
```diff
@@ -6,6 +6,7 @@ Requires=docker.socket
 
 [Service]
 ExecStart=/usr/bin/docker -d -H fd://
+MountFlags=slave
```
Comment (lsm5) [inline contrib/init/systemd/docker.service]: LGTM

**Comment by crosbymichael** [comment]
Y'all know that you can no longer go into `/var/lib/docker` from the host with this change and inspect/backup a containers filesystems because the mounts happen in a new namespace.  And by y'all i mean @vbatts ;)

**Comment by jfrazelle** [comment]
I am kinda in favor of possibly reverting this now because there are instances when people need to backup a stopped container etc.

**Comment by dqminh** [comment]
Being able to access current active mounts in `/var/lib/docker` is currently important for some of our use cases too ( we use aufs, but i guess that would apply to other drivers ), since we do have some tasks that runs outside of container and exposing the container's active filesystem.

**Comment by vbatts** [comment]
Wait now. You can still See the mountinfo of the pid. Give me an example of
what doesnt work. It will be _more_ beneficial to find a solution to your
problem than to continue running the docker daemon in the root mount
namespace.

**Comment by jfrazelle** [comment]
How would you view something in the `/var/lib/docker/aufs/mnt/{container_id}/` folder for example?

**Comment by jfrazelle** [comment]
Or backup a file from that directory if the container is stopped and can't be restarted

**Comment by dqminh** [comment]
@vbatts our usecase is roughly as follow, we have a process that chroots into `/var/lib/docker/aufs/mnt/{container_id}/` and accepts RPC requests to access/manipulate the mounted filesystem for that container.

I figure that we can `setns(docker_pid, mount namespace)` in addition to chroot when starting the process too so that it will be able access the mount namespace ?

**Comment by vbatts** [comment]
```
vbatts@noyee ~ (master) $ docker run -it -d busybox touch file.txt
f533ab8ecd3be89b79c4e6635a1562633110adb7b435f6b4c81813d4106e2681
vbatts@noyee ~ (master) $ sudo su -
[root@noyee ~]# cd /var/lib/docker/btrfs/subvolumes/f533ab8ecd3be89b79c4e6635a1562633110adb7b435f6b4c81813d4106e2681
[root@noyee f533ab8ecd3be89b79c4e6635a1562633110adb7b435f6b4c81813d4106e2681]# ls -l file.txt
-rw-r--r--. 1 root root 0 Jan 21 21:19 file.txt
```

and you can see that the daemon is in a different mount ns than everything else
```
[root@noyee ~]# ls -l /proc/*/ns/mnt | grep -A3 -B3 $(pgrep docker)
lrwxrwxrwx. 1 vbatts vbatts 0 Jan 21 21:10 /proc/2430/ns/mnt -> mnt:[4026531840]
lrwxrwxrwx. 1 root   root   0 Jan 21 21:10 /proc/24/ns/mnt -> mnt:[4026531840]
lrwxrwxrwx. 1 root   root   0 Jan 21 21:10 /proc/25/ns/mnt -> mnt:[4026531840]
lrwxrwxrwx. 1 root   root   0 Jan 21 21:10 /proc/26098/ns/mnt -> mnt:[4026532470]
lrwxrwxrwx. 1 vbatts vbatts 0 Jan 21 21:10 /proc/26292/ns/mnt -> mnt:[4026531840]
lrwxrwxrwx. 1 vbatts vbatts 0 Jan 21 21:10 /proc/26309/ns/mnt -> mnt:[4026531840]
lrwxrwxrwx. 1 vbatts vbatts 0 Jan 21 21:10 /proc/26310/ns/mnt -> mnt:[4026531840]
```

**Comment by vbatts** [comment]
just because the mount reference is not included in all other pid's mountinfo, does not mean the effects of the mounts are invisible.

**Comment by jfrazelle** [comment]
but thats btrfs, thats the problem I was testing on btrfs originally with this change and I didn't have the same problem I see with aufs and not being able to see things in the `mnt` dir

**Comment by jfrazelle** [comment]
:D

**Comment by jfrazelle** [comment]
does it burn :fire:

**Comment by jfrazelle** [comment]
I can reproduce with:
```
docker run -it --privileged dockercore/docker bash
```
then if you take that contianer id and try to ls /var/lib/docker/aufs/mnt/{container_id}/ nothing is there

**Comment by vbatts** [comment]
@dqminh `setns()` is the other thing I was going to give an example with.

**Comment by vbatts** [comment]
holy moly. i literally can't even. does AUFS even?

**Comment by vbatts** [comment]
works fine on overlayfs even. I don't know what to say for aufs. Really a shame that you'd improve many folks experience, reduce errors, but determine a need to revert because of mystical behavior on aufs.

**Comment by jfrazelle** [comment]
Is there another solution maybe :(

**Comment by vbatts** [comment]
@jfrazelle i even tried https://gist.github.com/vbatts/7257da4823c8bc8c61c1 and the directories are still hidden on aufs.

**Comment by jfrazelle** [comment]
*shakes fist in air* ok @crosbymichael you think we should revert?

**Comment by jfrazelle** [comment]
We could also add something to docs about how this recommended for
devicemapper

**Comment by dqminh** [comment]
I think the PR is safe as-is right ? It only modifies systemd / sysvinit-redhat init script and aufs is not generally used on those systems while devicemapper and btrfs are.

It's probably better to keep this and add some docs on how this doesnt work with aufs.

**Comment by jfrazelle** [comment]
I have debian aufs machines, actually the core machine is debian aufs.

**Comment by dqminh** [comment]
@jfrazelle debian should use sysvinit-debian script, which is not affected by this change right ?

**Comment by jfrazelle** [comment]
Not debian jessie ;)

**Comment by jfrazelle** [comment]
As much as I would love to be the only debian jessie user running aufs I do
not think that is the case

**Comment by jfrazelle** [comment]
Also what about arch linux users with aufs

**Comment by dqminh** [comment]
oh noes :fire:

**Comment by dqminh** [comment]
@vbatts weird, https://gist.github.com/vbatts/7257da4823c8bc8c61c1 works on my aufs system.

```
# without setns
> ls /var/lib/docker/aufs/mnt/f714499277a27743ad55e052d4ab1f7946a8baf98a64463e521aa9a0f25f242f

# with setns, 30411 is docker daemon pid
> gcc -o test-setns main.c
> sudo ./test-setns /proc/30411/ns/mnt ls /var/lib/docker/aufs/mnt/f714499277a27743ad55e052d4ab1f7946a8baf98a64463e521aa9a0f25f242f
bin  dev  etc  file.txt  home  lib  lib64  linuxrc  media  mnt  opt  proc  root  run  sbin  sys  tmp  usr  var
```

**Comment by vbatts** [comment]
@dqminh oh good. is that on a running or stopped container?

Perhaps this is the start of a contrib tool for debugging containers in their namespace.

**Comment by dqminh** [comment]
@vbatts it's on a running container. If the container is stopped, then `/var/lib/docker/aufs/mnt/container-id` should be empty right, because it's unmounted.

**Comment by philips** [comment]
@dqminh Now I am confused. Does your version of aufs differ from @jfrazelle 's?

**Comment by crosbymichael** [comment]
Running containers,  we don't leave them mounted if they are stopped

**Comment by vbatts** [comment]
#10281 is up for debate

**Comment by tianon** [comment]
@vbatts heh, looks like we forgot to push this over to sysvinit, upstart, and openrc too

**Comment by vbatts** [comment]
@tianon 

https://github.com/docker/docker/commit/6bb65864589fbd720622cbd795763d108999a366

is one, but the others were hackish, with `start-stop-daemon`. I may have a branch floating around here though...

**Comment by tianon** [comment]
Yeah, I remember that we discussed it quite a bit and thought we'd found a solution to the `start-stop-daemon` problems. :cry:

**Comment by tianon** [comment]
https://botbot.me/freenode/docker-maintainers/msg/30055754/

**Comment by vbatts** [comment]
"I want to die" that was the solution there

**Pull request merged**